src/Maestro/Microsoft.DotNet.Maestro.Tasks/src/PushMetadataToBuildAssetRegistry.cs

```diff
@@ -132,10 +133,29 @@ public async Task<bool> PushMetadataAsync(CancellationToken cancellationToken)
             return builds.Select(t => new BuildRef(t.Key, t.Value)).ToImmutableList();
         }
 
-        private async Task<int?> GetBuildId(DependencyDetail dep, IMaestroApi client, CancellationToken cancellationToken)
+        private static async Task<int?> GetBuildId(DependencyDetail dep, IMaestroApi client, Dictionary<(string name, string version), int> assetCache, CancellationToken cancellationToken)
         {
+            if (assetCache.TryGetValue((dep.Name, dep.Version), out int value))
+            {
+                return value;
+            }
             var assets = await client.Assets.ListAssetsAsync(name: dep.Name, version: dep.Version, cancellationToken: cancellationToken);
-            return assets.OrderByDescending(a => a.Id).FirstOrDefault()?.BuildId;
+            var buildId = assets.OrderByDescending(a => a.Id).FirstOrDefault()?.BuildId;
+            if (!buildId.HasValue)
+            {
+                return null;
+            }
+
+            var build = await client.Builds.GetBuildAsync(buildId.Value, cancellationToken);
+            foreach (var asset in build.Assets)
+            {
+                if (!assetCache.ContainsKey((asset.Name, asset.Version)))
+                {
+                    assetCache.Add((asset.Name, asset.Version), build.Id);
+                }
+            }
```

Comment:
Why grab all the assets of the build if no one is asking for them? That seems like a lot of work that no one asked us to do...

Reply:
The assumption is that if one asset in the Version.Details.xml is from a specific build, then many others will be also. This cache is useless if every thing in the file is from a different build.

aspnet/AspNetCore is an example, they have like 30 dependencies from the same build.

Reply:
Ill add a comment explaining this.

Reply:
Hrmm...  I guess it's a bit of a toss up. It adds a call sometimes, but presumably to reduce them later.

Let's do it!

---
In reply to: [271050956](https://github.com/dotnet/arcade-services/pull/296#discussion_r271050956) [](ancestors = 271050956)

Reply:
Yea, for a build that have 5 dependencies to different builds, its like 1 extra call per dependency that isn't that slow. For a build with 130 dependencies to 4 builds, its 8 total network calls, rather than 130.